### PR TITLE
more tests and more compatibility with dev-morevp branch

### PR DESCRIPTION
**Pull request by paulperegud**
Backports tests from dev-morevp branch. Also backports merkle tree with variable depth functionality. Adds some missing tests.

Depends on #16

### CODE REVIEW

**Comment by paulperegud** [comment]
TODO:
* how challengeExit gas cost is influenced by size of the queue
* behavior of gas costs for large queues and long operation (long running test)
* check if removing `require(...)` lines from code trips the tests

File: tests/contracts/root_chain/test_challenge_standard_exit.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import pytest
```
Comment (pgebal) [inline tests/contracts/root_chain/test_challenge_standard_exit.py]: Let's see what happens when someone tries to challenge finalized exit.

File: tests/contracts/root_chain/test_process_exits.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    assert testlang.get_standard_exit(spend_id).owner == owner.address
+    testlang.finalize_exits(NULL_ADDRESS)
+    testlang.forward_timestamp(1)
+    assert testlang.get_standard_exit(spend_id).owner == owner.address
+    testlang.finalize_exits(NULL_ADDRESS)
+    assert testlang.get_standard_exit(spend_id).owner == NULL_ADDRESS_HEX
+
+
+def test_finalize_exits_new_utxo_is_mature_after_mfp_pls_rep(testlang):
```
Comment (pgebal) [inline tests/contracts/root_chain/test_process_exits.py]: This test name is quite cryptic.
I think this test covers cases tested in `test_finalize_exits_old_utxo_is_mature_after_single_mfp`, so we can delete the previous test.

File: tests/contracts/root_chain/test_start_standard_exit.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+def test_start_standard_exit_twice_should_fail(testlang, utxo):
+    testlang.start_standard_exit(utxo.owner, utxo.spend_id)
+    with pytest.raises(TransactionFailed):
+        testlang.start_standard_exit(utxo.owner, utxo.spend_id)
+
+
+def test_start_standard_exit_invalid_proof_should_fail(testlang, utxo):
+    spend_tx = testlang.child_chain.get_transaction(utxo.spend_id)
+    with pytest.raises(TransactionFailed):
+        testlang.root_chain.startExit(utxo.spend_id, spend_tx.encoded, b'')
```
Comment (pgebal) [inline tests/contracts/root_chain/test_start_standard_exit.py]: RootChain.startExit takes 4 arguments. Here we provide just 3. How does that work?

File: tests/contracts/root_chain/test_start_standard_exit.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+import pytest
```
Comment (pgebal) [inline tests/contracts/root_chain/test_start_standard_exit.py]: I think we don't have a test for a failure of exiting utxo that is already exited.

File: tests/contracts/root_chain/test_deposit.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    deposit_id = testlang.deposit(owner, amount)
+    deposit_blknum, _, _ = decode_utxo_id(deposit_id)
+
+    plasma_block = testlang.get_plasma_block(deposit_blknum)
+    assert plasma_block.root == get_deposit_hash(address_to_bytes(owner.address), NULL_ADDRESS, amount)
+    assert plasma_block.timestamp == testlang.timestamp
+    assert testlang.root_chain.currentDepositBlock() == 2
+
+
+def test_deposit_zero_amount_should_succeed(testlang):
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply tests/contracts/root_chain/test_deposit.py]: We gain almost nothing by forbidding it. Also everyone else allows it. Principle of least astonishment.

File: tests/contracts/root_chain/test_fee_exit.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import pytest
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply tests/contracts/root_chain/test_fee_exit.py]: This is impossible, unfortunately. This is a type of failure where watchers detect and everyone exits before operator.

File: tests/contracts/root_chain/test_process_exits.py (hunk 2 above)
Comment (paulperegud) [reply tests/contracts/root_chain/test_process_exits.py]: They are similar and mechanically second one includes first one, but they declare (and test) two different properties.

File: tests/contracts/root_chain/test_exit_from_deposit.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import pytest
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply tests/contracts/root_chain/test_exit_from_deposit.py]: Added

File: contracts/ByteUtils.sol [hunk 8]
```diff
@@ -2,10 +2,21 @@ pragma solidity ^0.4.0;
 
 
 /**
- * @title Bytes operations
- * @dev Based on https://github.com/GNSPS/solidity-bytes-utils/blob/master/contracts/BytesLib.sol
+ * @title ByteUtils
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply contracts/ByteUtils.sol]: I'm treating those libraries as libraries - minimal changes.

File: contracts/ByteUtils.sol [hunk 9]
```diff
@@ -2,10 +2,21 @@ pragma solidity ^0.4.0;
...
+ * @dev Utilties for working with bytes.
  */
 library ByteUtils {
+    /*
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply contracts/ByteUtils.sol]: I'm treating those libraries as libraries - minimal changes.

File: plasma_core/utils/merkle/fixed_merkle.py [hunk 10]
```diff
@@ -1,76 +1,91 @@
...
         if not self.hashed:
             leaf = sha3(leaf)
         computed_hash = leaf
+        computed_index = index
+
         for i in range(0, self.depth * 32, 32):
-            segment = proof[i:i + 32]
-            if index % 2 == 0:
-                computed_hash = sha3(computed_hash + segment)
+            proof_segment = proof[i:i + 32]
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply plasma_core/utils/merkle/fixed_merkle.py]: done

File: tests/contracts/priority_queue/test_priority_queue.py [hunk 11]
```diff
@@ -52,3 +53,51 @@ def test_priority_queue_delete_then_insert(priority_queue):
     assert priority_queue.delMin() == 2
     priority_queue.insert(5)
     assert priority_queue.getMin() == 5
+
+
+def test_priority_queue_insert_spam_does_not_elevate_gas_cost_above_200k(priority_queue):
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply tests/contracts/priority_queue/test_priority_queue.py]: done

File: tests/contracts/priority_queue/test_priority_queue.py [hunk 12]
```diff
@@ -52,3 +53,51 @@ def test_priority_queue_delete_then_insert(priority_queue):
...
+    for i in range(1, len(values)):
+        if i % 10 == 0:
+            ethtester.chain.mine()
+        assert i == priority_queue.delMin()
+        gas = ethtester.chain.last_gas_used()
+        assert gas <= op_cost(len(values) - i)
+
+
+def op_cost(n):
+    return 21000 + 26538 + 6638 * math.floor(math.log(n, 2))
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply tests/contracts/priority_queue/test_priority_queue.py]: done

File: tests/contracts/root_chain/test_challenge_standard_exit.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+    testlang.start_standard_exit(owner, spend_id)
+    doublespend_id = testlang.spend_utxo(spend_id, owner, 100, owner)
+
+    testlang.challenge_standard_exit(spend_id, doublespend_id)
+
+    testlang.root_chain.exits(spend_id) == [address_to_hex(NULL_ADDRESS), 0]
+
+
+def test_challenge_standard_exit_invalid_spend_should_fail(ethtester, testlang):
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply tests/contracts/root_chain/test_challenge_standard_exit.py]: done

File: tests/contracts/root_chain/test_challenge_standard_exit.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    owner, amount = testlang.accounts[0], 100
+    deposit_id = testlang.deposit(owner.address, amount)
+    spend_id = testlang.spend_utxo(deposit_id, owner, 100, owner)
+
+    testlang.start_standard_exit(owner, spend_id)
+    doublespend_id = testlang.spend_utxo(spend_id, owner, 100, owner)
+
+    testlang.challenge_standard_exit(spend_id, doublespend_id)
+
+    testlang.root_chain.exits(spend_id) == [address_to_hex(NULL_ADDRESS), 0]
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply tests/contracts/root_chain/test_challenge_standard_exit.py]: fixed

File: tests/contracts/root_chain/test_deposit.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+        root_chain.addToken(token.address)
+
+
+def test_token_adding_eth_token_should_fail(root_chain):
+    assert root_chain.hasToken(NULL_ADDRESS)
+    with pytest.raises(TransactionFailed):
+        root_chain.addToken(NULL_ADDRESS)
+
+
+def test_token_deposit_should_succeed(testlang, root_chain, token):
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply tests/contracts/root_chain/test_deposit.py]: done

File: tests/contracts/root_chain/test_exit_from_deposit.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+    owner, amount = testlang.accounts[0], 100
+    deposit_id = testlang.deposit(owner, amount)
+
+    testlang.child_chain.get_transaction(deposit_id)
+
+    with pytest.raises(TransactionFailed):
+        testlang.root_chain.startDepositExit(deposit_id, NULL_ADDRESS_HEX, 1000, sender=owner.key)
+
+
+def test_start_deposit_exit_from_child_block_should_fail(testlang):
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply tests/contracts/root_chain/test_exit_from_deposit.py]: done

File: plasma_core/utils/utils.py [hunk 17]
```diff
@@ -19,4 +19,5 @@ def bytes_fill_left(inp, length):
 
 
 def get_deposit_hash(owner, token, value):
-    return u.sha3(owner + token + b'\x00' * 31 + u.int_to_bytes(value))
+    value_bytes = u.int_to_bytes(value) if value != 0 else b'\x00'
+    return u.sha3(owner + token + b'\x00' * 31 + value_bytes)
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply plasma_core/utils/utils.py]: done

File: testlang/testlang.py [hunk 18]
```diff
@@ -75,12 +75,25 @@ def __init__(self, root_chain, ethtester):
...
     def timestamp(self):
         """Current chain timestamp"""
         return self.ethtester.chain.head_state.timestamp
 
     def deposit(self, owner, amount):
+        if isinstance(owner, str):
+            for acc in self.accounts:
+                if acc.address == owner:
+                    owner = acc
+            if isinstance(owner, str):
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply testlang/testlang.py]: I've changed this, although in different manner. Should be way more readable.

File: plasma_core/utils/merkle/fixed_merkle.py [hunk 19]
```diff
@@ -1,76 +1,91 @@
...
+        self.left = left
+        self.right = right
 
 
 class FixedMerkle(object):
 
     def __init__(self, depth, leaves=[], hashed=False):
         if depth < 1:
-            raise ValueError('depth should be at least 1')
+            raise ValueError('depth must be at least 1')
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply plasma_core/utils/merkle/fixed_merkle.py]: punting to #23

**Pull request merged**